ocaml/xapi/xapi_pool.ml

```diff
@@ -554,57 +554,39 @@ let rec create_or_get_host_on_master __context rpc session_id (host_ref, host) :
 
   new_host_ref
 
-and create_or_get_sr_on_master __context rpc session_id (sr_ref, sr)
-  : API.ref_SR =
+and create_or_get_sr_on_master __context rpc session_id (sr_ref,sr): API.ref_SR =
   let my_uuid = sr.API.sR_uuid in
-  let new_sr_ref =
-    try Client.SR.get_by_uuid ~rpc ~session_id ~uuid:my_uuid with _ ->
-      let my_pbd_ref = List.hd (Db.SR.get_PBDs ~__context ~self:sr_ref) in
-      let my_pbd = Db.PBD.get_record ~__context ~self:my_pbd_ref in
-      let pbds_on_master = Client.PBD.get_all_records ~rpc ~session_id in
-      (* The only possible shared SRs are ISO, as other SRs cannot be shared properly across pools. *)
-      (* In this case, if we find a SR with a PBD having the same device_config field, we pick this SR instead of building a new one *)
-      let iso_already_exists_on_master () =
-        List.exists
-          (fun (_, x) ->
-            Listext.List.set_equiv x.API.pBD_device_config
-              my_pbd.API.pBD_device_config )
-          pbds_on_master
+  try
+    Client.SR.get_by_uuid ~rpc ~session_id ~uuid:my_uuid
+  with _ ->
+    if sr.API.sR_is_tools_sr then
+      (* find the tools ISO and return it *)
+      try
+        Client.SR.get_all_records ~rpc ~session_id
+        |> List.find (fun (_, sr) -> sr.API.sR_is_tools_sr)
+        |> (fun (ref,_) -> ref)
```

Comment:
Should we check that the device config is the same and what should happen if it isn't?
What would happen if you start a VM with the tools ISO plugged in, and migrate it to a host which has a newer tools SR? (e.g. during an RPU)
If the contents of the tools ISO is different then I think the SRs should have different UUIDs to avoid accidentally changing the contents of a plugged ISO in a guest.

Reply:
This is a good point. Should we scan for VMs that that a tools ISO attached?

Reply:
The host that is joining a pool cannot have any running VMs (another pre-join check further up in this file), so this should not be a problem. This function is not related to RPU either.